export types

turadg opened the pull request:
refs: #6343

## Description

Emit types for agoric-sdk package builds like https://github.com/endojs/endo/pull/1307

### Security Considerations

n/a

### Documentation Considerations

Export types are a documentation resource.

### Testing Considerations

CI. And dry release?

mhofman commented on packages/internal/src/utils.js:
```diff
@@ -139,15 +144,20 @@ export const applyLabelingError = (func, args, label = undefined) => {
 
 /**
- * @param {object} obj
+ * @param {Record<string | symbol, unknown>} obj
  * @returns {(string|symbol)[]}
```
Would be neat to have an `infer` based result here, but can be left as future improvement.

turadg replied:
it's correctly inferred now. I don't see the value in removing the `returns` but open to it.

mhofman replied:
I meant with some work we could type this utility function so that the it would actually have as return type the subset of the `obj` keys for which the values are functions. Aka a type implementation of the behavior.

Sure `(string|symbol)[]` is correct, but it's fairly broad.

turadg replied:
Ah got it. I think we'd run into this: https://stackoverflow.com/a/55012175/46040 (found from `https://github.com/microsoft/TypeScript/issues/41036`)

mhofman commented on packages/internal/src/utils.js:
```diff
@@ -241,6 +252,7 @@ export const deeplyFulfilledObject = obj => {
  * @returns {<T>(fn: () => Promise<T>) => Promise<{ result: T, duration: number }>}
  */
 export const makeMeasureSeconds = currentTimeMillisec => {
+  /** @param {() => any} fn */
```
For this one we should be able to easy propagate through the result by templetizing here
```suggestion
  /**
   * @template {() => any} T
   * @param {T} fn
   */

turadg replied:
What different will it make? `makeMeasureSeconds` defines the type already.

mhofman replied:
Right, sorry habit of letting inference flow to the return types and not use explicit `@returns`.

mhofman commented on packages/smart-wallet/src/walletFactory.js:
```diff
@@ -14,6 +14,11 @@ import { E } from '@endo/far';
 
+// Ambient types. Needed only for dev but this does a runtime import.
+import '@agoric/vats/exported.js';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import '@agoric/swingset-vat/src/types-ambient.js';
```
I don't suppose we can avoid relying on ambient types for `swingset-vat`. I know it does re-export most types explicitly at the top level.

mhofman commented on packages/SwingSet/src/kernel/state/kernelKeeper.js:
```diff
@@ -158,7 +158,7 @@ const FIRST_METER_ID = 1n;
 /**
  * @param {HostStore} hostStorage
- * @param {KernelSlog} kernelSlog
+ * @param {KernelSlog | null} kernelSlog
```
Oh I have a stash with this fix and other swingset ones somewhere

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -67,7 +68,7 @@ export const reserveThenDeposit = async (
   console.info('confirmed deposit for', debugName);
 };
 
-/** @type {<T>(store: ERef<MapStore>, key: string, make: () => T) => Promise<T>} */
+/** @type {<T>(store: ERef<Map<string, T>>, key: string, make: () => T) => Promise<T>} */
```
I'm surprised we have a `Map` based store around

turadg replied:
Good idea. Done.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -90,6 +91,7 @@ export const makeInstallCache = async (
   { installCacheKey = 'installCache', loadBundle },
 ) => {
   /** @type {CopyMap<string, {installation: Installation, boardId: string, path?: string}>} */
+  // @ts-expect-error cast
```
Please actually cast narrowly instead of ignoring all errors on the line.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -108,6 +110,7 @@ export const makeInstallCache = async (
   };
 
   const wrapInstall = install => async (mPath, bPath, opts) => {
+    // @ts-expect-error xxx bundle types
```
What is the issue here?

turadg replied:
> Property 'endoZipBase64Sha512' does not exist on type 'EndoZipBase64Bundle | GetExportBundle | NestedEvaluateBundle'.

mhofman replied:
That seems like a legitimate issue. Why suppress it?
Or is it because of an overload in `loadBundle` which isn't typed correctly?

turadg replied:
The type was wrong (which is what I meant by "xxx bundle types")

e6c5a934a5ab6af203151336c2eeb22aa096faf1 has a fix

mhofman replied:
I believe runtime changes are worth when they directly address issues found by typing. improvements

mhofman commented on packages/vats/src/repl.js:
```diff
@@ -97,6 +98,7 @@ function dump0(value, spaces, inProgress, depth) {
 
     const props = names
       .filter(n => nonNumber.has(n))
+      // @ts-expect-error
```
cast instead

mhofman commented on packages/vats/src/vat-provisioning.js:
```diff
@@ -50,6 +51,7 @@ export function buildRootObject(_vatPowers) {
         // Update the notifier when the chainBundle resolves.
         const { notifier, updater } = makeNotifierKit();
         chainBundle.then(clientHome => {
+          // @ts-expect-error IterationObserver not callable
           updater(harden({ clientHome, clientAddress: address }));
```
split in multiple lines and cast as appropriate?

turadg replied:
`updater` itself has the error,
> This expression is not callable.
>  Type 'IterationObserver<any>' has no call signatures.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -8,13 +8,13 @@ export const DEPOSIT_FACET = 'depositFacet';
 
 /**
- * @param {ERef<NameAdmin>} nameAdmin
+ * @param {ERef<import('@agoric/vats/src/nameHub.js').NameAdmin>} nameAdmin
```
How hard would it be to export types at the root of the `vats` package?

turadg replied:
good call. in the course of the PR they were exported but these imports needed updating.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -44,7 +44,7 @@ export const reserveThenGetNamePaths = async (nameAdmin, paths) => {
 
 /**
- * @param {ERef<NameAdmin>} nameAdmin
+ * @param {ERef<import('@agoric/vats/src/types.js').NameAdmin>} nameAdmin
```
Yeah definitely would like to avoid deep imports here.

mhofman commented on packages/vats/src/types.js:
```diff
@@ -1,5 +1,7 @@
 // @ts-check
 
+export {};
```
Looks like this is now exported types, so it should be easy to re-export from `index.js`

mhofman replied:
This file is currently still referenced by `vats/exported.js` using `import './src/types.js';`, which is no longer relevant.

Have we also confirmed there were no use of the typedefs in this file that may have silently rolled back to an implicit `any` type?

turadg replied:
> This file is currently still referenced by vats/exported.js using import './src/types.js';, which is no longer relevant.

Good catch. Removed.

> Have we also confirmed there were no use of the typedefs in this file that may have silently rolled back to an implicit any type?

Good idea. Found some (though I don't know if they were resolving before).

mhofman reviewed (commented):
Nice start. Not an approval until I can get some clarifications on the various `ts-expect-error`.

I'd like to at least double check the need for some of these `ts-expect-error`s. As I've expressed before, I believe they are worse than casts as they can hide multiple errors unexpectedly. If possible narrow application of casts are better (while slightly harder to read unfortunately).

Also I see there was an attempt to switch to exported types, but it was reverted. If attempted again, let's avoid deep imports cross package.

I have some type fixes for swingset that I may now try to brush off :)

mhofman commented:
Nit on reviewing process: While I understand this is still a draft PR, it's extremely difficult to track changes compared to the previously reviewed state if force pushes include both amendments and base changes. In general please avoid do not amend and stick to commits additions once reviews have started and until the PR is accepted, even if those are marked as `fixup` to indicate such amendments to be made just before merging only. Furthermore when changing the base, please do 2 force push: one push adding commits on the old base as is, and one force push rebasing showing no changes (or only conflict resolution if warrented).

Another point specific to this PR: please include the main package name affected in the conventional commit prefix, e.g. `chore(ERTP)`. Right now there are commits with duplicate names making it impossible to understand what's going on.

I really want type improvements, but right now this PR is extremely difficult to review.

turadg commented:
I agree with the amendment policy for PRs marked R4R. I don't PRs that aren't. This is something our eng org should agree on so let's discuss in standup.

> Another point specific to this PR: please include the main package name affected in the conventional commit prefix, e.g. chore(ERTP).

I've opted against that because IIUC it creates redundant info in READMEs. e.g. the `@agoric/ertp` changelog says "chore(ertp): foo".

> Right now there are commits with duplicate names making it impossible to understand what's going on.

I'm confused on why omission of package name makes it impossible to understand what's going on. If I had not separated the commits and instead, for example, squashed all the `chore: misc types fixes` then there would be no duplicates but I expect it would be harder to follow the changes.

> I really want type improvements, but right now this PR is extremely difficult to review.

All the changes since your review of the draft were addressing your comments. So maybe it suffices to look over your comments and see whether they were addressed to your satisfaction.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -90,8 +91,12 @@ export const makeInstallCache = async (
-    makeCopyMap([]),
+
+  const initial = await provideWhen(
+    // @ts-expect-error cast
+    E.get(homeP).scratch,
```
While I mentioned above I'd prefer to figure out the `provideWhen`'s param type to the compatible subtype, is there a reason to use `// @ts-expect-error` for casts instead of doing an actual cast ?

Do we consider casting to be too visually onerous?

turadg replied:
Moot with 570cc993f3cfb2eeff49a363fd10b497b100079c, but I do find casting in JSdoc to be hard to read. No strong feeling one way or the other.

mhofman commented on packages/vats/src/vat-provisioning.js:
```diff
@@ -50,7 +51,7 @@ export function buildRootObject(_vatPowers) {
         const { notifier, updater } = makeNotifierKit();
         chainBundle.then(clientHome => {
-          updater(harden({ clientHome, clientAddress: address }));
+          updater.updateState(harden({ clientHome, clientAddress: address }));
```
Fix LGTM, but I should probably defer to someone more familiar with this code and notifiers. I believe this is the only behavioral difference in this PR.

turadg replied:
Yep, the only intended behavior change. `updater` comes out of `makeNotifierKit` which just has `updateState`, `fail` and `finish`. The comment says "Update the notifier" so I'm sure it's not meant to be one of the latter two.

FYI @michaelfig who I think wrote this.

mhofman commented on packages/vats/src/core/boot.js:
```diff
@@ -17,6 +17,13 @@ import * as behaviors from './behaviors.js';
+// Ambient types
+import './types.js';
+// eslint-disable-next-line import/export -- no named exports
+export * from '../types.js';
```
I just realized this file is the entrypoint of the package. It's definitely a little confusing.

Would you mind introducing an `index.js` at the root which `exports *` from both this file and `./src/types.js` ?

turadg replied:
Good idea

mhofman reviewed (commented):
I'd like to do another pass, but here is my review so far.

I still have some concerns regarding casts and type compatibility.

I believe the conversion of the `vats` package to exported type introduces some mistaken re-exports, and could use a better layering for the main entrypoint.

turadg commented on packages/zoe/src/internal-types.js:
```diff
@@ -241,7 +241,7 @@
  * @property {ShutdownWithFailure} terminateWithFailure
  * Terminate the vat in which the contract is running as a failure.
- * @property {(bundleCap: BundleCap, options?: Record<string, any>) => Promise<RootAndAdminNode>} upgrade
+ * @property {(bundleCap: *, options?: Record<string, any>) => Promise<RootAndAdminNode>} upgrade
```
It's a type in name only.
https://github.com/Agoric/agoric-sdk/blob/8980a2d57e54d635f1d85b9d80b62ca5046b4f43/packages/SwingSet/src/types-external.js#L353

This was expedient to work around an import.

turadg commented on packages/smart-wallet/package.json:
```diff
@@ -32,6 +34,7 @@
     "@agoric/nat": "^4.1.0",
     "@agoric/notifier": "^0.5.1",
     "@agoric/store": "^0.8.3",
+    "@agoric/swingset-vat": "^0.30.2",
```
Ticket here https://github.com/Agoric/agoric-sdk/issues/6512

turadg commented on packages/SwingSet/tools/internal-types.js:
```diff
@@ -1,13 +0,0 @@
-/**
- * @typedef {object} ManualTimerAdmin
- * @property { (when: Timestamp) => void } advanceTo
- */
-
-/**
- * @typedef {ManualTimerAdmin & TimerService} ManualTimer
- */
```
I believe the differences were tech debt.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -68,7 +68,7 @@ export const reserveThenDeposit = async (
-/** @type {<T>(store: ERef<Map<string, T>>, key: string, make: () => T) => Promise<T>} */
+/** @type {<T>(store: ERef<Pick<MapStore<string, T>, 'get' | 'set'>>, key: string, make: () => T) => Promise<T>} */
 const provideWhen = async (store, key, make) => {
   const found = await E(store).get(key);
```
I probably should have looked at the actual implementation before.

I was under the impression that `MapStore` threw when doing a `get` of an uninitialized value. That's what I meant earlier by different semantics, unfortunately the signature of `get` methods only varies on the return type between `Map` and `MapStore`, making them "compatible". Thinking more about this, it's probably a footgun that these objects have such a similar shape, yet such different semantics.

Anyway, I think `provideWhen` is faulty because of this, and I don't understand how the fallback path would have every worked with the `MapStore` variant.

turadg replied:
Agree with your assessment. Is there a requested change in there? An issue to make or a thread to start?

mhofman replied:
I don't know. I don't think it'd be easy enough to fix now, but at least we should create an issue?

I think reverting the `Pick` and adding a `ts-expect-error` pointing to the issue?

turadg replied:
I made an issue but then realized a simpler fix: 713298

mhofman replied:
I fail to see how it's a fix?

mhofman commented on packages/SwingSet/src/types-external.js:
```diff
@@ -351,7 +351,7 @@ export {};
  * @typedef { string } BundleID
  * @typedef {*} BundleCap
- * @typedef { { moduleFormat: 'endoZipBase64', endoZipBase64: string, endoZipBase64Sha512 } } EndoZipBase64Bundle
+ * @typedef { { moduleFormat: 'endoZipBase64', endoZipBase64: string, endoZipBase64Sha512: string } } EndoZipBase64Bundle
```
Oh yeah that would do it!

mhofman commented on packages/vats/index.js:
```diff
@@ -0,0 +1,9 @@
+export * from './src/core/boot.js';
+// eslint-disable-next-line import/export -- no named exports
+export * from './src/types.js';
+export * from './src/nameHub.js';
+export * from './src/bridge.js';
```
I still believe this will cause formerly internal runtime helpers to now be visible on the root of this package, not only types.

See https://github.com/Agoric/agoric-sdk/pull/6500#discussion_r1008709131

turadg replied:
Thanks for the thorough review!

> I suspect the ./bridge.js import is for the BridgeManager type? In that case I believe the better approach would be to have a JSDoc @typedef in ../types.js and leverage the existing re-export from there. Same the nameHub.js's NameHub and NameAdmin.

`makeNameHubAdmin` is imported in three other modules, so arguably this top level is an improvement, but I'll err towards conservative and leave runtime exports refactoring for another day.

mhofman commented on packages/inter-protocol/src/proposals/utils.js:
```diff
@@ -81,7 +81,7 @@ const provideWhen = async (store, key, make) => {
 /**
  *
- * @param {{ scratch: ERef<MapStore<string, CopyTagged>> }} homeP
+ * @param {{ scratch: ERef<Map<string, CopyTagged>> }} homeP
```
Does `scratch` actually have `Map` semantics? If not that'd be a very unobvious cast that hide the runtime differences.

turadg replied:
Yes, it does https://github.com/Agoric/agoric-sdk/blob/bf9b05ffef83d26a98df53953ee231bca637ae5b/packages/solo/src/scratch.js#L17-L20

mhofman commented on packages/SwingSet/src/vats/plugin-manager.js:
```diff
@@ -21,9 +21,11 @@ const DEFAULT_WALKER = Far('walker', { walk: pluginRootP => pluginRootP });
 /**
  * @template T
- * @typedef {T} Device
+ * @typedef {'Device' & { __deviceType__: T }} Device
```
🎉

mhofman commented on packages/agoric-cli/src/init.js:
```diff
@@ -2,8 +2,10 @@ import chalk from 'chalk';
 // Ambient types. Needed only for dev but this does a runtime import.
+// https://github.com/Agoric/agoric-sdk/issues/6512
 import '@endo/captp/src/types.js';
 import '@agoric/swingset-vat/exported.js';
+import '@agoric/swingset-vat/src/vats/network/types.js';
```
Seeing this now, would it make sense to add `import './src/vats/network/types.js';` to `@agoric/swingset-vat/exported.js` instead ?

Edit: maybe not since it's used directly in other places?

turadg replied:
yeah, I considered that and opted for lesser change to what's exported. I figure it will be addressed in further ambients work

mhofman reviewed (commented):
I'm not sure I understand how the `provideWhen` type change is actually a fix.

mhofman reviewed (approved):
LGTM

turadg commented:
> I would have like to see more coordination around getting rid of @ts-check. It's a useful explicit tag, especially when moving code between repositories.

@dckc it's an easy change to reverse. You want every JS file in a package to have `@ts-check` even if it's redundant with the package configuration?

dckc commented:
> You want every JS file in a package to have @ts-check even if it's redundant with the package configuration?

That would have been my preference, yes. At this point, I don't feel strongly that it's worth more churn in the git history. But it seems like a change to house style that's worth fairly widespread coordination.
